src/driver_zwave.cpp

```diff
@@ -0,0 +1,452 @@
+  if (zconnection_send_async(panConnection_, binaryCommand, binaryCommandLen,
+                             0)) {
+    panConnectionBusy_ = true;
+  }
+}
+
+void ZWaveDriver::AddNode() {
+  net_mgmt_learn_mode_start();
+
+  int idx = 0;
+  static uint8_t buf[200];
+
+  const uint8_t COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION = 0x34;
+  const uint8_t NODE_ADD = 0x01;
+
+  buf[idx++] = COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION;
+  buf[idx++] = NODE_ADD;
+  buf[idx++] = get_unique_seq_no();
+  buf[idx++] = 0;
+  buf[idx++] = 0x07; /* ADD_NODE_S2 */
+  buf[idx++] = 0;    /* Normal power, no NWI */
+
+  zconnection_send_async(gwZipconnection_, buf, idx, 0);
+}
+
+void ZWaveDriver::RemoveNode() {
+  int idx = 0;
+  static uint8_t buf[200];
+
+  const uint8_t COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION = 0x34;
+  const uint8_t NODE_REMOVE = 0x03;
+
+  buf[idx++] = COMMAND_CLASS_NETWORK_MANAGEMENT_INCLUSION;
+  buf[idx++] = NODE_REMOVE;
+  buf[idx++] = get_unique_seq_no();
+  buf[idx++] = 0;
+  buf[idx++] = 0x01; /* REMOVE_NODE_ANY */
+
+  zconnection_send_async(gwZipconnection_, buf, idx, 0);
+}
+
+void ZWaveDriver::SetDefault() {
+  int idx = 0;
+  static uint8_t buf[200];
+
+  const uint8_t COMMAND_CLASS_NETWORK_MANAGEMENT_BASIC = 0x4D;
+  const uint8_t DEFAULT_SET = 0x06;
+
+  buf[idx++] = COMMAND_CLASS_NETWORK_MANAGEMENT_BASIC;
+  buf[idx++] = DEFAULT_SET;
+  buf[idx++] = get_unique_seq_no();
+
+  zconnection_send_async(gwZipconnection_, buf, idx, 0);
+}
+
+void ZWaveDriver::List() {
+  ZwaveMsg msg;
+
+  std::cout << "List of discovered Z/IP services:" << std::endl;
+  for (zip_service* n = zresource_get(); n; n = n->next) {
+    std::cout << " host_name:    " << n->host_name << std::endl;
+    std::cout << " service_name: " << n->service_name << std::endl;
+    std::cout << " infolen: " << std::dec << n->infolen << std::endl;
+    std::cout << " info: " << std::endl;
+
+    ZwaveMsg_ZwaveNode* node = msg.add_node();
+    node->set_service_name(n->service_name);
+
+    for (int i = 0; i < n->infolen; i++) {
+      if (ZwaveClassType_IsValid(n->info[i])) {
+        const std::string& className =
+            ZwaveClassType_Name(static_cast<ZwaveClassType>(n->info[i]));
+
+        ZwaveClassType zwave_class;
+        ZwaveClassType_Parse(className, &zwave_class);
+
+        node->add_zwave_class(zwave_class);
+
+        std::cout << "  " << std::hex << (int)n->info[i] << " " << className
+                  << std::endl;
+      } else {
+        std::cout << "  " << std::hex << static_cast<int>(n->info[i])
+                  << " not found in the ZwaveClassType enum." << std::endl;
+      }
+    }
+    std::cout << std::endl;
+  }
+  msg.set_result(true);
+
+  std::string buffer;
+  msg.SerializeToString(&buffer);
+  zqm_push_update_->Send(buffer);
+}
+
+bool ZWaveDriver::ConnectToGateway() {
+  gwZipconnection_ = ZipConnect(serverIP_.c_str());
+
+  if (gwZipconnection_) return true;
```

Comment:
can gwZipconnection_ be null? Of so, compare to nullptr. Othersize just return gwZipconnection_;